File: src/app/core/services/job.service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import { Job } from '../models/job.model';
import { ProfileService } from './profile.service';
import { MatchEngineService, MatchBreakdown } from './match-engine.service';

export interface JobWithMatch extends Job {
  match: MatchBreakdown;
}

@Injectable({
  providedIn: 'root'
})
export class JobService {
  private profileService = inject(ProfileService);
  private matchEngineService = inject(MatchEngineService);

  private jobsSignal = signal<Job[]>([]);
  readonly jobs = this.jobsSignal.asReadonly();

  private readonly SEED_JOBS: Job[] = [
    {
      id: 'job-stripe',
      title: 'Frontend Engineer - Payments UI',
      company: 'Stripe',
      companyLogo: 'https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?w=128&h=128&fit=crop&crop=faces&q=80',
      description: 'Join the payments flow team to design, build, and optimize Stripe Elements and Checkout interfaces. You will work heavily with Angular, TypeScript, and fine-tune CSS layouts to ensure seamless, accessible payment checkout experiences for millions of consumers.',
      location: 'Chicago, IL',
      workMode: 'hybrid',
      gpaThreshold: 3.5,
      skillsRequired: ['Angular', 'TypeScript', 'SCSS', 'RxJS', 'Web Performance'],
      sponsorshipRequired: false,
      salaryRange: '$135,000 - $165,000',
      postedDate: '2026-07-15',
      category: 'Frontend Engineering'
    },
    {
      id: 'job-vercel',
      title: 'Angular Developer (Internship)',
      company: 'Vercel',
      companyLogo: 'https://images.unsplash.com/photo-1618005198143-e5283b519a7f?w=128&h=128&fit=crop&crop=faces&q=80',
      description: 'We are looking for an Angular Intern who is passionate about build systems, fast rendering, and building beautiful, responsive dashboards. You will help build and maintain our integration portals, connecting enterprise frameworks with the Vercel platform.',
      location: 'Remote',
      workMode: 'remote',
      gpaThreshold: 3.2,
      skillsRequired: ['Angular', 'TypeScript', 'SCSS', 'Git'],
      sponsorshipRequired: true,
      salaryRange: '$45 - $60 / hour',
      postedDate: '2026-07-16',
      category: 'Frontend Engineering'
    },
    {
      id: 'job-linear',
      title: 'UI Developer & Product Designer',
      company: 'Linear',
      companyLogo: 'https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?w=128&h=128&fit=crop&crop=faces&q=80',
      description: 'At Linear, we care deeply about visual precision, speed, and clean interactions. You will span the gap between design and frontend code. Working in our UI platform team, you will implement smooth, high-fidelity components, micro-animations, and keyboard shortcuts.',
      location: 'New York, NY',
      workMode: 'onsite',
      gpaThreshold: 3.0,
      skillsRequired: ['TypeScript', 'SCSS', 'Animations', 'Figma', 'Angular'],
      sponsorshipRequired: false,
      salaryRange: '$140,000 - $175,000',
      postedDate: '2026-07-14',
      category: 'Design & Engineering'
    },
    {
      id: 'job-google',
      title: 'Software Engineer - Spring Boot Platform',
      company: 'Google',
      companyLogo: 'https://images.unsplash.com/photo-1572021335469-31706a17aaef?w=128&h=128&fit=crop&crop=faces&q=80',
      description: 'Develop back-end platform components for cloud applications. In this role, you will build microservices using Spring Boot, manage relational databases (SQL), write REST APIs, and optimize Docker deployments running on Google Cloud Platform.',
      location: 'Mountain View, CA',
      workMode: 'hybrid',
      gpaThreshold: 3.6,
      skillsRequired: ['Java', 'Spring Boot', 'SQL', 'REST API', 'Docker'],
      sponsorshipRequired: true,
      salaryRange: '$165,000 - $210,000',
      postedDate: '2026-07-12',
      category: 'Backend Engineering'
    },
    {
      id: 'job-notion',
      title: 'Full Stack Engineer - Collaboration Tools',
      company: 'Notion',
      companyLogo: 'https://images.unsplash.com/photo-1560250097-0b93528c311a?w=128&h=128&fit=crop&crop=faces&q=80',
      description: 'Work across the stack to build rich collaborative workspace documents. You will integrate frontend Angular standalone views with backend Spring Boot services. You should be comfortable with reactive programming in RxJS and database normalization in SQL.',
      location: 'San Francisco, CA',
      workMode: 'hybrid',
      gpaThreshold: 3.4,
      skillsRequired: ['TypeScript', 'Angular', 'Java', 'Spring Boot', 'SQL'],
      sponsorshipRequired: true,
      salaryRange: '$130,000 - $160,000',
      postedDate: '2026-07-17',
      category: 'Full Stack Engineering'
    },
    {
      id: 'job-airbnb',
      title: 'Junior UI Engineer',
      company: 'Airbnb',
      companyLogo: 'https://images.unsplash.com/photo-1568602471122-7832951cc4c5?w=128&h=128&fit=crop&crop=faces&q=80',
      description: 'Help us redefine modern digital travel bookings. We are seeking a Junior UI Engineer to focus on component libraries, responsive grids, and clean design patterns. You will build and test modules in Angular, collaborating with Figma designers.',
      location: 'Remote',
      workMode: 'remote',
      gpaThreshold: 3.0,
      skillsRequired: ['Angular', 'TypeScript', 'SCSS', 'RxJS'],
      sponsorshipRequired: false,
      salaryRange: '$95,000 - $120,000',
      postedDate: '2026-07-10',
      category: 'Frontend Engineering'
    }
  ];

  // Live computed property that returns jobs with calculated match scores for the active profile,
  // sorted by match score descending.
  readonly matchedJobs = computed<JobWithMatch[]>(() => {
    const profile = this.profileService.currentProfile();
    const jobs = this.jobsSignal();
    
    return jobs.map(job => {
      const match = this.matchEngineService.calculateMatch(profile, job);
      return {
        ...job,
        match
      };
    }).sort((a, b) => b.match.score - a.match.score);
  });

  private bookmarkedIdsSignal = signal<string[]>([]);
  readonly bookmarkedIds = this.bookmarkedIdsSignal.asReadonly();

  constructor() {
    this.initJobs();
    this.initBookmarks();
  }

  private initBookmarks() {
    const cached = localStorage.getItem('credx_bookmarks');
    if (cached) {
      try {
        this.bookmarkedIdsSignal.set(JSON.parse(cached));
      } catch (e) {
        localStorage.setItem('credx_bookmarks', JSON.stringify([]));
      }
    }
  }

  toggleBookmark(jobId: string) {
    this.bookmarkedIdsSignal.update(current => {
      const updated = current.includes(jobId)
        ? current.filter(id => id !== jobId)
        : [...current, jobId];
      localStorage.setItem('credx_bookmarks', JSON.stringify(updated));
      return updated;
    });
  }

  isBookmarked(jobId: string): boolean {
    return this.bookmarkedIdsSignal().includes(jobId);
  }

  private initJobs() {
    const cached = localStorage.getItem('credx_jobs');
    if (cached) {
      try {
        this.jobsSignal.set(JSON.parse(cached));
      } catch (e) {
        this.jobsSignal.set(this.SEED_JOBS);
        localStorage.setItem('credx_jobs', JSON.stringify(this.SEED_JOBS));
      }
    } else {
      this.jobsSignal.set(this.SEED_JOBS);
      localStorage.setItem('credx_jobs', JSON.stringify(this.SEED_JOBS));
    }
  }

  getJobById(id: string): JobWithMatch | null {
    const jobs = this.matchedJobs();
    return jobs.find(j => j.id === id) || null;
  }

  addJob(jobData: Omit<Job, 'id' | 'postedDate'>) {
    const newJob: Job = {
      ...jobData,
      id: `job-${Math.random().toString(36).substring(2, 9)}`,
      postedDate: new Date().toISOString().split('T')[0]
    };

    this.jobsSignal.update(current => {
      const updated = [newJob, ...current];
      localStorage.setItem('credx_jobs', JSON.stringify(updated));
      return updated;
    });
  }
}

File: backend/src/main/java/com/credx/match/service/JobService.java
package com.credx.match.service;

import com.credx.match.dto.JobDTO;
import com.credx.match.dto.JobMatchDTO;
import com.credx.match.dto.StudentProfileDTO;
import com.credx.match.entity.Job;
import com.credx.match.entity.Skill;
import com.credx.match.repository.JobRepository;
import com.credx.match.repository.SkillRepository;
import jakarta.annotation.PostConstruct;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;
import java.util.stream.Collectors;

@Service
public class JobService {

    private final JobRepository jobRepository;
    private final SkillRepository skillRepository;
    private final MatchEngineService matchEngineService;

    public JobService(JobRepository jobRepository, SkillRepository skillRepository, MatchEngineService matchEngineService) {
        this.jobRepository = jobRepository;
        this.skillRepository = skillRepository;
        this.matchEngineService = matchEngineService;
    }

    @PostConstruct
    public void initSeedJobs() {
        if (jobRepository.count() == 0) {
            seedInitialJobs();
        }
    }

    public List<JobDTO> getAllJobs() {
        return jobRepository.findAll().stream().map(this::convertToDTO).collect(Collectors.toList());
    }

    public JobDTO getJobById(Long id) {
        Job job = jobRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Job posting not found with ID: " + id));
        return convertToDTO(job);
    }

    public List<JobMatchDTO> getRecommendedJobs(StudentProfileDTO student) {
        List<Job> jobs = jobRepository.findAll();
        List<JobMatchDTO> matches = new ArrayList<>();

        for (Job job : jobs) {
            JobDTO jobDTO = convertToDTO(job);
            matches.add(matchEngineService.calculateMatch(student, jobDTO));
        }

        // Sort descending by score
        matches.sort((a, b) -> Integer.compare(b.getScore(), a.getScore()));
        return matches;
    }

    public JobMatchDTO getJobMatchForStudent(Long jobId, StudentProfileDTO student) {
        JobDTO jobDTO = getJobById(jobId);
        return matchEngineService.calculateMatch(student, jobDTO);
    }

    @Transactional
    public JobDTO createJob(JobDTO dto) {
        Job job = new Job();
        job.setTitle(dto.getTitle());
        job.setCompany(dto.getCompany());
        job.setCompanyLogo(dto.getCompanyLogo() != null ? dto.getCompanyLogo() : "https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?w=128&h=128&fit=crop&q=80");
        job.setDescription(dto.getDescription());
        job.setLocation(dto.getLocation());
        job.setWorkMode(dto.getWorkMode() != null ? dto.getWorkMode() : "remote");
        job.setGpaThreshold(dto.getGpaThreshold() != null ? dto.getGpaThreshold() : 3.0);
        job.setSponsorshipRequired(dto.getSponsorshipRequired() != null ? dto.getSponsorshipRequired() : false);
        job.setSalaryRange(dto.getSalaryRange() != null ? dto.getSalaryRange() : "$90,000 - $120,000");
        job.setPostedDate("Just now");
        job.setCategory(dto.getCategory() != null ? dto.getCategory() : "Frontend Engineering");

        if (dto.getSkillsRequired() != null) {
            Set<Skill> skillEntities = new HashSet<>();
            for (String skillName : dto.getSkillsRequired()) {
                Skill skill = skillRepository.findByName(skillName)
                        .orElseGet(() -> skillRepository.save(new Skill(skillName)));
                skillEntities.add(skill);
            }
            job.setSkillsRequired(skillEntities);
        }

        Job saved = jobRepository.save(job);
        return convertToDTO(saved);
    }

    private JobDTO convertToDTO(Job job) {
        JobDTO dto = new JobDTO();
        dto.setId(job.getId().toString());
        dto.setTitle(job.getTitle());
        dto.setCompany(job.getCompany());
        dto.setCompanyLogo(job.getCompanyLogo());
        dto.setDescription(job.getDescription());
        dto.setLocation(job.getLocation());
        dto.setWorkMode(job.getWorkMode());
        dto.setGpaThreshold(job.getGpaThreshold());
        dto.setSponsorshipRequired(job.getSponsorshipRequired());
        dto.setSalaryRange(job.getSalaryRange());
        dto.setPostedDate(job.getPostedDate());
        dto.setCategory(job.getCategory());
        dto.setSkillsRequired(job.getSkillsRequired().stream().map(Skill::getName).collect(Collectors.toList()));
        return dto;
    }

    private void seedInitialJobs() {
        createSeedJob("job-stripe", "Frontend Engineer - Payments UI", "Stripe", "Chicago, IL (Hybrid)", "hybrid", 3.5, Arrays.asList("Angular", "TypeScript", "SCSS", "RxJS"), true, "$110,000 - $135,000", "2 days ago", "Frontend Engineering", "https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?w=128&h=128&fit=crop&q=80");
        createSeedJob("job-vercel", "Junior Angular Developer", "Vercel", "Remote", "remote", 3.2, Arrays.asList("Angular", "TypeScript", "SCSS", "Figma"), true, "$95,000 - $115,000", "1 day ago", "Frontend Engineering", "https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?w=128&h=128&fit=crop&q=80");
        createSeedJob("job-google", "Software Engineer - Spring & Cloud", "Google", "Mountain View, CA", "onsite", 3.7, Arrays.asList("Java", "Spring Boot", "SQL", "Docker"), false, "$130,000 - $160,000", "3 days ago", "Backend Engineering", "https://images.unsplash.com/photo-1573804633927-bfcbcd909acd?w=128&h=128&fit=crop&q=80");
        createSeedJob("job-linear", "Full Stack Web Intern", "Linear", "Remote", "remote", 3.0, Arrays.asList("TypeScript", "Angular", "SQL", "REST API"), true, "$45 / hour", "4 days ago", "Full Stack", "https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?w=128&h=128&fit=crop&q=80");
        createSeedJob("job-notion", "UI/UX Developer", "Notion", "San Francisco, CA", "hybrid", 3.4, Arrays.asList("TypeScript", "SCSS", "Figma", "Angular"), true, "$105,000 - $130,000", "5 days ago", "UI/UX Engineering", "https://images.unsplash.com/photo-1568602471122-7832951cc4c5?w=128&h=128&fit=crop&q=80");
        createSeedJob("job-airbnb", "Junior UI Engineer", "Airbnb", "Chicago, IL", "hybrid", 3.3, Arrays.asList("Angular", "TypeScript", "SCSS", "RxJS"), true, "$100,000 - $125,000", "6 days ago", "Frontend Engineering", "https://images.unsplash.com/photo-1568602471122-7832951cc4c5?w=128&h=128&fit=crop&q=80");
        createSeedJob("job-datadog", "Backend Microservices Associate", "Datadog", "New York, NY", "onsite", 3.5, Arrays.asList("Java", "Spring Boot", "Docker", "SQL"), false, "$120,000 - $145,000", "1 week ago", "Backend Engineering", "https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?w=128&h=128&fit=crop&q=80");
        createSeedJob("job-figma", "Design Systems Specialist", "Figma", "Remote", "remote", 3.2, Arrays.asList("Figma", "TypeScript", "SCSS", "Angular"), true, "$115,000 - $140,000", "1 week ago", "UI/UX Engineering", "https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?w=128&h=128&fit=crop&q=80");
    }

    private void createSeedJob(String idStr, String title, String company, String location, String mode, double gpa, List<String> skills, boolean sponsor, String salary, String date, String cat, String logo) {
        Job job = new Job();
        job.setTitle(title);
        job.setCompany(company);
        job.setCompanyLogo(logo);
        job.setDescription("Detailed role description for " + title + " at " + company + ". Join a high-growth team building modern software.");
        job.setLocation(location);
        job.setWorkMode(mode);
        job.setGpaThreshold(gpa);
        job.setSponsorshipRequired(sponsor);
        job.setSalaryRange(salary);
        job.setPostedDate(date);
        job.setCategory(cat);

        Set<Skill> skillEntities = new HashSet<>();
        for (String sName : skills) {
            Skill skill = skillRepository.findByName(sName)
                    .orElseGet(() -> skillRepository.save(new Skill(sName)));
            skillEntities.add(skill);
        }
        job.setSkillsRequired(skillEntities);

        jobRepository.save(job);
    }
}
